### PR TITLE
Issue #3187: TIMESTAMPTZ <=> VARCHAR

### PR DESCRIPTION
**Pull request by hawkfish**
Add string casts that understand time zones.
These are now used for display, so any test that uses them 
needs to set the calendar and time zone explicitly.

Made some tweaks and fixed some minor bugs
 in the default parsing and formatting code.

### CODE REVIEW

File: extension/icu/icu-strptime.cpp [hunk 1]
```diff
@@ -157,6 +159,77 @@ struct ICUStrptime : public ICUDateFunc {
...
+		const auto tz_name = info->tz_setting.c_str();
+
+		UnaryExecutor::ExecuteWithNulls<string_t, timestamp_t>(
+		    source, result, count, [&](string_t input, ValidityMask &mask, idx_t idx) {
+			    timestamp_t result;
+			    const auto str = input.GetDataUnsafe();
+			    const auto len = input.GetSize();
+			    string_t tz(nullptr, 0);
+			    if (!Timestamp::TryConvertTimestampTZ(str, len, result, tz)) {
+				    auto msg = Timestamp::ConversionError(string(str, len));
```
Comment (Mytherin) [inline extension/icu/icu-strptime.cpp:176]: Perhaps use `HandleCastError::AssignError` here instead of reimplementing the branching/throwing?

**Review by Mytherin — COMMENTED** [review]
Thanks for the PR! LGTM. One minor comment.

On a related note, would it perhaps also make sense for `NOW` to return a `TIMESTAMPTZ`?

**Comment by hawkfish** [comment]
> On a related note, would it perhaps also make sense for `NOW` to return a `TIMESTAMPTZ`?

Yeah I was wondering about that too.

**Comment by Mytherin** [comment]
There seems to be a missing `DUCKDB_API` on the `BoundCastInfo` constructor, leading to the Windows failure:

```
2022-10-07T07:01:15.6628516Z icu-strptime.obj : error LNK2019: unresolved external symbol "public: __cdecl duckdb::BoundCastInfo::BoundCastInfo(bool (__cdecl*)(class duckdb::Vector &,class duckdb::Vector &,unsigned __int64,struct duckdb::CastParameters &),class std::unique_ptr<struct duckdb::BoundCastData,struct std::default_delete<struct duckdb::BoundCastData> >)" (??0BoundCastInfo@duckdb@@QEAA@P6A_NAEAVVector@1@0_KAEAUCastParameters@1@@ZV?$unique_ptr@UBoundCastData@duckdb@@U?$default_delete@UBoundCastData@duckdb@@@std@@@std@@@Z) referenced in function "public: static struct duckdb::BoundCastInfo __cdecl duckdb::ICUStrptime::BindCastFromVarchar(struct duckdb::BindCastInput &,struct duckdb::LogicalType const &,struct duckdb::LogicalType const &)" (?BindCastFromVarchar@ICUStrptime@duckdb@@SA?AUBoundCastInfo@2@AEAUBindCastInput@2@AEBULogicalType@2@1@Z) [D:\a\duckdb\duckdb\extension\icu\icu_loadable_extension.vcxproj]
2022-10-07T07:01:15.7018113Z D:\a\duckdb\duckdb\extension\icu\icu.duckdb_extension : fatal error LNK1120: 1 unresolved externals [D:\a\duckdb\duckdb\extension\icu\icu_loadable_extension.vcxproj]

```

**Pull request merged**